### ISSUE TITLE
Alternatives are very slow for LM/AStar+turn costs

### ISSUE THREAD

**Issue by easbar** [graphhopper/graphhopper#2566]
map: bayern-210101.osm.pbf: 
route using maps 2 client: http://localhost:3000/?point=49.583276%2C10.991626&point=48.455783%2C12.095752&profile=car
profile: fastest/car

![image](https://user-images.githubusercontent.com/17603532/165449184-f4af0de6-54c1-4926-af20-836849e0827b.png)

|traversal mode|single route|alternatives (max=3)|
|-|-|-|
|node-based|13ms|~800ms|
|edge-based|25ms|~9000ms|

So alternatives take 80 times longer than a single route for node-based. This seems very slow, but at least it still takes less than a second. But for edge-based it is even worse where the slow down is x360 and the response takes so long that its almost not useful anymore.

This is not just special for this route, I just gave an example. I actually saw routes that took tens of seconds, and I did not even try a larger map. Disabling LM makes it even slower.

**Comment by easbar** [comment]
Using the approach taken by @michaz in #2092 the query time for edge-based alternative seems to go down from ~9s to ~5s. However, for node-based the response becomes (slightly) slower, like ~950ms and also the single routes for edge-based seem to take longer (~40ms vs. ~25ms), but this was just a quick manual check.

**Comment by easbar** [comment]
@otbutz in #2567  you wrote:

> I can vouch for [Java Mission Control](https://adoptium.net/de/jmc/) paired with -XX:+UnlockDiagnosticVMOptions -XX:+DebugNonSafepoints for getting a good grasp about where the bottleneck is :)

I am answering here as this is about speeding up alternative routes:

Thanks I will try. I am using the YourKit Profiler, usually. But in my experience this does not help much and the real issues are typically caused by the algorithms (i.e. normally something is called too often rather than some method is slow). In the case of AlternativeRoute I am pretty sure this takes very long, especially for large search trees:

https://github.com/graphhopper/graphhopper/blob/47febfc6278e6eebf15f2940e5bac0c77c0034b5/core/src/main/java/com/graphhopper/routing/AlternativeRoute.java#L321-L326

But I did not really look at the alternative route algorithm to figure out if this can be avoided without making the quality of alternative routes worse.

**Comment by otbutz** [comment]
> normally something is called too often rather than some method is slow

The flame view is immensely helpful IMHO.

https://www.youtube.com/watch?v=vVh_z34Gtn8

**Comment by easbar** [comment]
Yes, I like the flame view, too. YourKit has it as well. But often I saw cases where something looks like a bottleneck there, but in reality it wasn't. Overall I feel like I spent a lot of time looking at the profiler, but rarely it helped me finding important bottlenecks and a lot more often improving the algorithm lead to significant speed ups. But maybe I'm still just doing it wrong.

In other news: This: https://github.com/graphhopper/graphhopper/commit/19e2e16bfa9ff663846f99113bd17d23b12be487 is definitely an improvement, especially for edge-based routing without CH. And in this case the PriorityQueue#remove call actually shows up in the profiler, too :) I'll open a separate PR for this soon.
